cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -493,6 +567,22 @@ class DefinitionByReferenceNode extends InstructionNode {
     not exists(instr.getPrimaryInstruction().(CallInstruction).getStaticCallTarget()) and
     result = "output argument"
   }
+
+  override Function getFunction() { result = instr.getEnclosingFunction() }
+
+  override IRType getType() { result = instr.getResultIRType() }
+
+  override Location getLocation() { result = instr.getLocation() }
+
+  // Make the read side effect's side effect operand the pre update node of this write side effect.
+  // This ensures that we match up the parameter index of the parameter indirection's modification.
+  override Node getPreUpdateNode() {
+    exists(ReadSideEffectInstruction read |
+      read.getPrimaryInstruction() = instr.getPrimaryInstruction() and
+      read.getArgumentDef() = instr.getDestinationAddress() and
```

Comment:
This looks like a fundamentally difficult/impossible formula to join-order.

Reply:
I think the join order produced looks reasonable. Here's the relevant tuple counts on ChakreCore:
```
Tuple counts for DataFlowUtil::PostUpdateNode::getPreUpdateNode_dispred#ff/2@33fe9f:
385895  ~4%     {2} r1 = SCAN DataFlowUtil::DefinitionByReferenceNode#class#ff AS I OUTPUT I.<1>, I.<0> 'this'
385564  ~0%     {3} r2 = JOIN r1 WITH Instruction::WriteSideEffectInstruction::getDestinationAddress_dispred#ff AS R ON FIRST 1 OUTPUT R.<1>, r1.<1> 'this', r1.<0>
373778  ~0%     {3} r3 = JOIN r2 WITH Instruction::ReadSideEffectInstruction::getArgumentDef_dispred#ff_10#join_rhs AS R ON FIRST 1 OUTPUT R.<1>, r2.<1> 'this', r2.<2>
373778  ~1%     {4} r4 = JOIN r3 WITH Instruction::ReadSideEffectInstruction::getSideEffectOperand_dispred#ff AS R ON FIRST 1 OUTPUT R.<1>, r3.<1> 'this', r3.<2>, r3.<0>
373778  ~7%     {4} r5 = JOIN r4 WITH DataFlowUtil::TOperandNode#ff@staged_ext AS R ON FIRST 1 OUTPUT r4.<3>, r4.<1> 'this', r4.<2>, R.<1> 'result'
373778  ~0%     {4} r6 = JOIN r5 WITH Instruction::SideEffectInstruction::getPrimaryInstruction_dispred#3#ff AS R ON FIRST 1 OUTPUT r5.<2>, R.<1>, r5.<1> 'this', r5.<3> 'result'
373756  ~0%     {2} r7 = JOIN r6 WITH Instruction::SideEffectInstruction::getPrimaryInstruction_dispred#3#ff AS R ON FIRST 2 OUTPUT r6.<2> 'this', r6.<3> 'result'
545664  ~0%     {2} r8 = DataFlowUtil::PartialDefinitionNode::getPreUpdateNode_dispred#ff \/ r7
                return r8
```

(I think the union at the end is from another `getPreUpdateNode` implementation)

Reply:
Looks good. No concerns, then.